### PR TITLE
Identify and mitigate weaknesses

### PR DESCRIPTION
**Pull request by lucasgonze**
Fixes https://github.com/eBay/NuRaft/issues/424

# Description

Used OpenSSF Security Scorecard to identify and mitigate weaknesses in [NuRaft](https://github.com/eBay/NuRaft). Raised the project's score from 5.5 to 10. See branch https://github.com/lucasgonze/NuRaft/pull/new/pin-dependencies

# Changes

1. Pinned dependencies in cmake.yml. This prevents attacks via upstream actions. See https://github.com/ossf/scorecard/blob/e42af756609b2cde6d757fd45ea05ddf0016ff62/docs/checks.md#pinned-dependencies

2. Set token permissions to read-only in cmake.yml. This follows the principle of least privilege. This is important because attackers may use a compromised token with write access to push malicious code into the project.

3. Moved inclusion of CodeCov for code coverage metrics to a Github Action from github_build_action.sh and travis_build.sh. This enables dependency pinning on the action and setting token permissions to read-only for least privilege.

4. Enable Dependabot version updates for GitHub Actions. Configure Dependabot via dependabot.yml file to enable security checks. See https://github.com/ossf/scorecard/blob/e42af756609b2cde6d757fd45ea05ddf0016ff62/docs/checks.md#dependency-update-tool

Note: Dependabot is unable to track vulnerabilities in the Asio library. Implementors should periodically check and, if necessary, manually upgrade the patch level using `git submodule update --recursive --remote`. The installation instructions offer alternative ways to incorporate Asio, including a prepare.sh script that clones it without a submodule. This should be deprecated except in the build.

Todo: create a pull request. It would be helpful to get an introduction to the NuRaft maintainers from Justin or Ian.

# Scorecard Output - This PR

```
{
  "date": "2023-04-08",
  "repo": {
    "name": "file://.",
    "commit": "no commits found"
  },
  "scorecard": {
    "version": "4.4.0",
    "commit": "e42af756609b2cde6d757fd45ea05ddf0016ff62"
  },
  "score": 10,
  "checks": [
    {
      "details": null,
      "score": 10,
      "reason": "no binaries found in the repo",
      "name": "Binary-Artifacts",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/e42af756609b2cde6d757fd45ea05ddf0016ff62/docs/checks.md#binary-artifacts",
        "short": "Determines if the project has generated executable (binary) artifacts in the source repository."
      }
    },
    {
      "details": null,
      "score": 10,
      "reason": "no dangerous workflow patterns detected",
      "name": "Dangerous-Workflow",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/e42af756609b2cde6d757fd45ea05ddf0016ff62/docs/checks.md#dangerous-workflow",
        "short": "Determines if the project's GitHub Action workflows avoid dangerous patterns."
      }
    },
    {
      "details": [
        "Info: Dependabot detected: .github/dependabot.yml:1"
      ],
      "score": 10,
      "reason": "update tool detected",
      "name": "Dependency-Update-Tool",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/e42af756609b2cde6d757fd45ea05ddf0016ff62/docs/checks.md#dependency-update-tool",
        "short": "Determines if the project uses a dependency update tool."
      }
    },
    {
      "details": [
        "Info: : LICENSE:1"
      ],
      "score": 10,
      "reason": "license file detected",
      "name": "License",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/e42af756609b2cde6d757fd45ea05ddf0016ff62/docs/checks.md#license",
        "short": "Determines if the project has defined a license."
      }
    },
    {
      "details": [
        "Info: GitHub-owned GitHubActions are pinned",
        "Info: Third-party GitHubActions are pinned",
        "Info: Dockerfile dependencies are pinned",
        "Info: no insecure (not pinned by hash) dependency downloads found in Dockerfiles",
        "Info: no insecure (not pinned by hash) dependency downloads found in shell scripts"
      ],
      "score": 10,
      "reason": "all dependencies are pinned",
      "name": "Pinned-Dependencies",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/e42af756609b2cde6d757fd45ea05ddf0016ff62/docs/checks.md#pinned-dependencies",
        "short": "Determines if the project has declared and pinned its dependencies."
      }
    },
    {
      "details": [
        "Info: top level 'contents' permission set to 'read': .github/workflows/cmake.yml:22"
      ],
      "score": 10,
      "reason": "tokens are read-only in GitHub workflows",
      "name": "Token-Permissions",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/e42af756609b2cde6d757fd45ea05ddf0016ff62/docs/checks.md#token-permissions",
        "short": "Determines if the project's workflows follow the principle of least privilege."
      }
    }
  ],
  "metadata": null
}

```

# Scorecard Output - Previous

```
{
  "date": "2023-04-26T13:23:49-07:00",
  "repo": {
    "name": "github.com/ebay/NuRaft",
    "commit": "2e4856bd2d23f6c6461a070c7d13fd49dab287d5"
  },
  "scorecard": {
    "version": "4.10.5",
    "commit": "27cfe92ed356fdb5a398c919ad480817ea907808"
  },
  "score": 5.2,
  "checks": [
    {
      "details": null,
      "score": 10,
      "reason": "no binaries found in the repo",
      "name": "Binary-Artifacts",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#binary-artifacts",
        "short": "Determines if the project has generated executable (binary) artifacts in the source repository."
      }
    },
    {
      "details": null,
      "score": 1,
      "reason": "branch protection is not maximal on development and all release branches",
      "name": "Branch-Protection",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#branch-protection",
        "short": "Determines if the default and release branches are protected with GitHub's branch protection settings."
      }
    },
    {
      "details": null,
      "score": 10,
      "reason": "30 out of 30 merged PRs checked by a CI test -- score normalized to 10",
      "name": "CI-Tests",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#ci-tests",
        "short": "Determines if the project runs tests before pull requests are merged."
      }
    },
    {
      "details": null,
      "score": 0,
      "reason": "no badge detected",
      "name": "CII-Best-Practices",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#cii-best-practices",
        "short": "Determines if the project has an OpenSSF (formerly CII) Best Practices Badge."
      }
    },
    {
      "details": null,
      "score": 0,
      "reason": "found 13 unreviewed human changesets",
      "name": "Code-Review",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#code-review",
        "short": "Determines if the project requires code review before pull requests (aka merge requests) are merged."
      }
    },
    {
      "details": null,
      "score": 10,
      "reason": "6 different organizations found -- score normalized to 10",
      "name": "Contributors",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#contributors",
        "short": "Determines if the project has a set of contributors from multiple organizations (e.g., companies)."
      }
    },
    {
      "details": null,
      "score": 10,
      "reason": "no dangerous workflow patterns detected",
      "name": "Dangerous-Workflow",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#dangerous-workflow",
        "short": "Determines if the project's GitHub Action workflows avoid dangerous patterns."
      }
    },
    {
      "details": null,
      "score": 0,
      "reason": "no update tool detected",
      "name": "Dependency-Update-Tool",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#dependency-update-tool",
        "short": "Determines if the project uses a dependency update tool."
      }
    },
    {
      "details": null,
      "score": 0,
      "reason": "project is not fuzzed",
      "name": "Fuzzing",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#fuzzing",
        "short": "Determines if the project uses fuzzing."
      }
    },
    {
      "details": null,
      "score": 10,
      "reason": "license file detected",
      "name": "License",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#license",
        "short": "Determines if the project has defined a license."
      }
    },
    {
      "details": null,
      "score": 10,
      "reason": "20 commit(s) out of 30 and 0 issue activity out of 30 found in the last 90 days -- score normalized to 10",
      "name": "Maintained",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#maintained",
        "short": "Determines if the project is \"actively maintained\"."
      }
    },
    {
      "details": null,
      "score": -1,
      "reason": "no published package detected",
      "name": "Packaging",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#packaging",
        "short": "Determines if the project is published as a package that others can easily download, install, easily update, and uninstall."
      }
    },
    {
      "details": null,
      "score": 4,
      "reason": "dependency not pinned by hash detected -- score normalized to 4",
      "name": "Pinned-Dependencies",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#pinned-dependencies",
        "short": "Determines if the project has declared and pinned the dependencies of its build process."
      }
    },
    {
      "details": null,
      "score": 0,
      "reason": "SAST tool is not run on all commits -- score normalized to 0",
      "name": "SAST",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#sast",
        "short": "Determines if the project uses static code analysis."
      }
    },
    {
      "details": null,
      "score": 10,
      "reason": "security policy file detected",
      "name": "Security-Policy",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#security-policy",
        "short": "Determines if the project has published a security policy."
      }
    },
    {
      "details": null,
      "score": -1,
      "reason": "no releases found",
      "name": "Signed-Releases",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#signed-releases",
        "short": "Determines if the project cryptographically signs release artifacts."
      }
    },
    {
      "details": null,
      "score": 0,
      "reason": "non read-only tokens detected in GitHub workflows",
      "name": "Token-Permissions",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#token-permissions",
        "short": "Determines if the project's workflows follow the principle of least privilege."
      }
    },
    {
      "details": null,
      "score": 10,
      "reason": "no vulnerabilities detected",
      "name": "Vulnerabilities",
      "documentation": {
        "url": "https://github.com/ossf/scorecard/blob/27cfe92ed356fdb5a398c919ad480817ea907808/docs/checks.md#vulnerabilities",
        "short": "Determines if the project has open, known unfixed vulnerabilities."
      }
    }
  ],
  "metadata": null
}

```

### CODE REVIEW

File: github_action_build.sh [hunk 1]
```diff
@@ -15,7 +15,6 @@ echo "branch name: ${BRANCH_NAME}"
 if [ -z "${GITHUB_HEAD_REF}" ]; then
     cmake -DCMAKE_BUILD_TYPE=Debug -DCODE_COVERAGE=1 -DENABLE_RAFT_STATS=1 ../
     make -j2 raft_cov
-    bash <(curl -s https://codecov.io/bash) -f raft_cov.info.cleaned
```
Comment (greensky00) [inline github_action_build.sh:18]: Hi @lucasgonze , thanks for submitting this PR.

I understand all the other changes you made, but I don't get how things work without this line; as `codecov-action` does not know what the result file is. 

This project uses `lcov`, it generates `raft_cov.info` and then `raft_cov.info.cleaned` after excluding given paths. Shouldn't we pass some parameters to `codecov-action`?

Comment (greensky00) [reply github_action_build.sh:18]: Maybe I should test it in other temporary repository.

**Comment by greensky00** [comment]
I got this error as expected
https://github.com/greensky00/NuRaft/actions/runs/4944335648/jobs/8839699515#step:5:54
but looks like `-f` flag can manually assign the result file.

**Comment by greensky00** [comment]
@lucasgonze 
Confirmed this additional parameter makes it work. 
https://github.com/greensky00/NuRaft/commit/6c46e5cb9eb1758322856ce52c6dacadd4dc5782
Please let me know if you want me to do it.

**Comment by lucasgonze** [comment]
> Please let me know if you want me to do it.

Confirmed.

**Review by greensky00 — APPROVED** [review]

**Pull request merged**